src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs

```diff
@@ -1,124 +1,241 @@
+        minimumAmountLabel.Visible = false;
+
+        switch ((TargetType)index)
         {
-            colour.Color = shownChoices[index].Colour;
+            case TargetType.Compound:
+                compounds.Visible = true;
+                compoundLabel.Visible = true;
+                minimumAmount.Visible = true;
+                minimumAmountLabel.Visible = true;
+                break;
+            case TargetType.Species:
+                species.Visible = true;
+                speciesLabel.Visible = true;
+                break;
         }
+
+        ApplySelectionColour();
+    }
+
+    public Vector2 GetMinDialogSize()
+    {
+        return new Vector2(400, 320);
     }
 
     protected override void Dispose(bool disposing)
     {
         if (disposing)
         {
-            if (CompoundsPath != null)
+            if (TargetTypesPath != null)
             {
+                TargetTypesPath.Dispose();
                 CompoundsPath.Dispose();
+                CompoundsLabelPath.Dispose();
+                SpeciesPath.Dispose();
+                SpeciesLabelPath.Dispose();
                 MaximumDistancePath.Dispose();
                 MinimumAmountPath.Dispose();
+                MinimumAmountLabelPath.Dispose();
                 ColourPath.Dispose();
             }
         }
 
         base.Dispose(disposing);
     }
+
+    /// <summary>
+    ///   Sets the GUI up to reflect an existing configuration
+    /// </summary>
+    private void LoadConfiguration(ChemoreceptorUpgrades configuration,
+        List<Compound> shownCompoundChoices, List<Species> shownSpeciesChoices)
+    {
+        if (configuration.TargetCompound != null)
+        {
+            TypeChanged((int)TargetType.Compound);
+            targetTypes.Selected = (int)TargetType.Compound;
+            compounds.Selected = shownCompoundChoices.FindIndex(c => c == configuration.TargetCompound);
+        }
+        else if (configuration.TargetSpecies != null)
+        {
+            TypeChanged((int)TargetType.Species);
+            targetTypes.Selected = (int)TargetType.Species;
+            species.Selected = shownSpeciesChoices.FindIndex(c => c == configuration.TargetSpecies);
+        }
+
+        maximumDistance.Value = configuration.SearchRange;
+        minimumAmount.Value = configuration.SearchAmount;
+        colour.Color = configuration.LineColour;
+    }
+
+    /// <summary>
+    ///   Applies the color of a selected target to the color picker
+    /// </summary>
+    private void ApplySelectionColour()
+    {
+        if (targetTypes.Selected == (int)TargetType.Compound && shownCompoundChoices != null
+            && compounds.Selected >= 0)
+            colour.Color = shownCompoundChoices[compounds.Selected].Colour;
+        if (targetTypes.Selected == (int)TargetType.Species && shownSpeciesChoices != null
+            && species.Selected >= 0)
+            colour.Color = shownSpeciesChoices[species.Selected].Colour;
```

Comment:
```suggestion
        if (targetTypes.Selected == (int)TargetType.Compound && shownCompoundChoices != null
            && compounds.Selected >= 0)
        {
            colour.Color = shownCompoundChoices[compounds.Selected].Colour;
        }

        if (targetTypes.Selected == (int)TargetType.Species && shownSpeciesChoices != null
            && species.Selected >= 0)
        {
            colour.Color = shownSpeciesChoices[species.Selected].Colour;
        }
```

Multiline conditions should use braces for readability purposes.